dali/dfu/dfuutil.cpp

```diff
@@ -857,21 +857,61 @@ class CDFUhelper: public CInterface, implements IDFUhelper
             else
                 throwError1(DFUERR_DFileNotFound, lfn);
         }
-        else {
-            Owned<IFileDescriptor> fdesc = file->getFileDescriptor();
-            Owned<IPropertyTree> t = fdesc->getFileTree();
+        else
+        {
+            Owned<IPropertyTree> t = queryDistributedFileDirectory().getFileTree(lfn, user);
             toXML(t, out, true);
         }
         return out;
     }
 
-    void addFileXML(const char *lfn,const StringBuffer &xml,IUserDescriptor *user)
+    void addFileXML(const char *lfn, const StringBuffer &xml, IUserDescriptor *user)
     {
         Owned<IPropertyTree> t = createPTreeFromXMLString(xml);
-        Owned<IFileDescriptor> fdesc = deserializeFileDescriptorTree(t,&queryNamedGroupStore(),0);
-        Owned<IDistributedFile> file = queryDistributedFileDirectory().createNew(fdesc,true);
-        if (file)
-            file->attach(lfn,user);
+        IDistributedFileDirectory &dfd = queryDistributedFileDirectory();
+        // Check if this XML is a superfile map
+        if (t->hasProp("@numsubfiles"))
+        {
+            if (dfd.exists(lfn, user))
+                throw MakeStringException(-1, "Super file '%s' already exists!", lfn);
+
+            // It seems XML is fine to create  super file
+            Owned<IDistributedSuperFile> superfile = dfd.createNewSuperFile(t);
+
+            superfile->validate();
+
+            try
+            {
+                superfile->attach(lfn, user, true);
```

Comment:
Attaching a super (calling super->attach()) will always involve linking the subfiles, so having a different attach() prototype is not necessary afaics.
i.e. a single overloaded version of > void attach(const char *_logicalname,IUserDescriptor *user) 
is all that is required. It must load + link the subfiles (see previous comments above)